engine/pkg/client/executor_group.go

```diff
@@ -0,0 +1,209 @@
+	"go.uber.org/zap"
+)
+
+const (
+	getExecutorBlockTimeout = 10 * time.Second
+)
+
+// ExecutorGroup holds a group of ExecutorClients.
+type ExecutorGroup interface {
+	// GetExecutorClient tries to get the ExecutorClient for the given executor.
+	GetExecutorClient(id model.ExecutorID) (ExecutorClient, bool)
+
+	// GetExecutorClientB tries to get the ExecutorClient for the given executor.
+	// It blocks until either the context has been canceled or the executor ID becomes valid.
+	GetExecutorClientB(ctx context.Context, id model.ExecutorID) (ExecutorClient, error)
+}
+
+// DefaultExecutorGroup is the default implementation for ExecutorGroup.
+type DefaultExecutorGroup struct {
+	mu      sync.RWMutex
+	clients map[model.ExecutorID]ExecutorClient
+
+	logger        *zap.Logger
+	clientFactory executorClientFactory
+}
+
+// NewExecutorGroup creates a new ExecutorGroup.
+func NewExecutorGroup(
+	credentials *security.Credential,
+	logger *zap.Logger,
+) *DefaultExecutorGroup {
+	return newExecutorGroupWithClientFactory(
+		logger,
+		newExecutorClientFactory(credentials, logger))
+}
+
+func newExecutorGroupWithClientFactory(
+	logger *zap.Logger,
+	factory executorClientFactory,
+) *DefaultExecutorGroup {
+	if logger == nil {
+		logger = zap.L()
+	}
+	return &DefaultExecutorGroup{
+		clients:       make(map[model.ExecutorID]ExecutorClient),
+		clientFactory: factory,
+		logger:        logger,
+	}
+}
+
+// GetExecutorClient tries to get the ExecutorClient for the given executor.
+func (g *DefaultExecutorGroup) GetExecutorClient(id model.ExecutorID) (ExecutorClient, bool) {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	if client, ok := g.clients[id]; ok {
+		return client, true
+	}
+	return nil, false
+}
+
+// GetExecutorClientB tries to get the ExecutorClient for the given executor.
+// It blocks until either the context has been canceled or the executor ID becomes valid.
+func (g *DefaultExecutorGroup) GetExecutorClientB(ctx context.Context, id model.ExecutorID) (ExecutorClient, error) {
+	ctx, cancel := context.WithTimeout(ctx, getExecutorBlockTimeout)
+	defer cancel()
+
+	var ret ExecutorClient
+	err := retry.Do(ctx, func() error {
+		if client, ok := g.GetExecutorClient(id); ok {
+			ret = client
+			return nil
+		}
+		return ErrExecutorNotFound.GenWithStack(&ExecutorNotFoundErrInfo{ExecutorID: id})
+	},
+		retry.WithBackoffMaxDelay(10*time.Second.Milliseconds()),
+		retry.WithIsRetryableErr(func(err error) bool {
+			return ErrExecutorNotFound.Is(err)
+		}))
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// UpdateExecutorList updates the stored clients using a map from executor IDs
+// to their addresses.
+// Note: This method will not wait for the clients to be fully connected.
+// In the rare case where grpc.Dial does fail, an error will be returned to the caller
+// of this method, and the caller should retry appropriately.
+func (g *DefaultExecutorGroup) UpdateExecutorList(executors map[model.ExecutorID]string) error {
```

Comment:
Where will this API be used?

Reply:
There are no existing use cases for now, but I'm imagining that this can be useful for some optimizations such as periodically re-fetching the list of executors, etc.